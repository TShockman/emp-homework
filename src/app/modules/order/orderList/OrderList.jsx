import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import Loading from '../../common/loading';
import OrderPreview from '../orderPreview';
import style from './orderList.scss';

// Component to display the current user's list of orders. Is
// fully connected by './OrderListContainer.js' and should be imported
// from './index.js'
// NOTE: in retrospect it might have be better to have this component
// take orders as a prop from it's parent, rather than have it's container
// directly pull it out of the store. This would have decoupled the order
// list from the single-unprivileged-user-details case and allowed for
// easier implementation of an admin view or other uses of a user list.
// If this were the case, the componentDidMount function would not be necessary.
export default class OrderList extends PureComponent {
  static propTypes = {
    orders: PropTypes.array,
    id: PropTypes.number,
    requestUserOrders: PropTypes.func.isRequired,
    push: PropTypes.func.isRequired
  };

  // redirect user to the login page if they are not logged in,
  // otherwise request the user's orders
  componentDidMount = () => {
    const {id, requestUserOrders, push} = this.props;
    if (id === null) {
      return push('/login');
    }
    return requestUserOrders({userId: id});
  };

  render() {
    const {orders} = this.props;

    // show loading animation if we don't have the orders yet
    if (!orders) {
      return <Loading/>;
    }

    return (
        <div className="album">
          <div className="row">
          {orders.map(order => <div className="col-md-6" key={order.id}><OrderPreview order={order}/></div>)}
          </div>
        </div>
    );
  }
}