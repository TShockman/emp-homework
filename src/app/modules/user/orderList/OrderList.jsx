import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import Loading from '../../common/loading';
import {ListGroup, ListGroupItem} from 'reactstrap';
import OrderPreview from '../orderPreview'

export default class OrderList extends PureComponent {
  static propTypes = {
    orders: PropTypes.array,
    id: PropTypes.number,
    router: PropTypes.object.isRequired,
    requestUserOrders: PropTypes.func.isRequired,
    push: PropTypes.func.isRequired
  };

  componentDidMount = () => {
    const {orders, id, router, requestUserOrders, push} = this.props;
    const {id: routerId} = router.params;

    if (id === null) {
      return push('/login');
    } else if (id.toString() !== routerId) {
      console.log('Can only view orders of user logged in.');
      // TODO: handle this differently?
      return push(`/user/${id}/orders`);
    } else {
      return requestUserOrders({id: router.params.id});
    }
  }

  render() {
    const {orders} = this.props;
    if (!orders) {
      return <Loading/>;
    }
    console.log(orders)

    return (
        <ListGroup>
          {orders.map(order => <ListGroupItem key={order.id}><OrderPreview order={order}/></ListGroupItem>)}
        </ListGroup>
    );
  }
}