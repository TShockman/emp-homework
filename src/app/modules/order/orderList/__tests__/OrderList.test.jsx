import React from 'react';
import ReactShallowRenderer from 'react-test-renderer/shallow';
import OrderList from '../OrderList';
import {mount} from 'enzyme';

const renderer = new ReactShallowRenderer();

describe('OrderList', () => {
  it('Consistently renders loading component when data has not arrived yet', () => {
    const props = {
      orders: undefined,
      id: 111,
      requestUserOrders: jest.fn(),
      push: jest.fn()
    };
    expect(renderer.render(<OrderList {...props}/>)).toMatchSnapshot();
  });
  it('Consistently renders with non-empty orders list', () => {
    const props = {
      orders: [
          {
            id: 123,
            ref: "mockref123-456",
            status: "ok",
            items: [{sku: "123-123-123", name: "Test Product", amount: 120}],
          },
          {
            id: 456,
            ref: "mockref908-765",
            status: "ok",
            items: [{sku: "123-123-123", name: "Test Product", amount: 120}],
          }],
      id: 111,
      requestUserOrders: jest.fn(),
      push: jest.fn()
    };
    expect(renderer.render(<OrderList {...props}/>)).toMatchSnapshot();
  });
  it('Redirects on mount if user is not logged in', () => {
    const props = {
      orders: undefined,
      id: null,
      requestUserOrders: jest.fn(),
      push: jest.fn()
    };
    const comp = <OrderList {...props}/>;
    expect(props.push).not.toHaveBeenCalled();

    mount(comp);
    expect(props.push).toHaveBeenLastCalledWith("/login");
  });
  it('Fetches user orders on mount if user is logged in', () => {
    const props = {
      orders: undefined,
      id: 123,
      requestUserOrders: jest.fn(),
      push: jest.fn()
    };
    const comp = <OrderList {...props}/>;
    expect(props.requestUserOrders).not.toHaveBeenCalled();

    mount(comp);
    expect(props.requestUserOrders).toHaveBeenLastCalledWith({userId: 123});
  });
});