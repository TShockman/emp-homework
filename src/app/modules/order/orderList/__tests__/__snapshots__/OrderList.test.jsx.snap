// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderList Consistently renders loading component when data has not arrived yet 1`] = `<Loading />`;

exports[`OrderList Consistently renders with non-empty orders list 1`] = `
<div
  className="album"
>
  <div
    className="row"
  >
    <div
      className="col-md-6"
    >
      <OrderPreview
        order={
          Object {
            "id": 123,
            "items": Array [
              Object {
                "amount": 120,
                "name": "Test Product",
                "sku": "123-123-123",
              },
            ],
            "ref": "mockref123-456",
            "status": "ok",
          }
        }
      />
    </div>
    <div
      className="col-md-6"
    >
      <OrderPreview
        order={
          Object {
            "id": 456,
            "items": Array [
              Object {
                "amount": 120,
                "name": "Test Product",
                "sku": "123-123-123",
              },
            ],
            "ref": "mockref908-765",
            "status": "ok",
          }
        }
      />
    </div>
  </div>
</div>
`;
